dali/pipeline/operator/direct_operator.h

```diff
@@ -0,0 +1,241 @@
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+      ThreadPool *tp) {
+    DALI_FAIL("Unsupported backends in DirectOperator.Run() with thread pool.");
+  }
+
+  // Runs operator using specified CUDA stream.
+  template <typename InBackend, typename OutBackend>
+  DLL_PUBLIC std::vector<std::shared_ptr<TensorList<OutBackend>>> Run(
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+      cudaStream_t cuda_stream) {
+    DALI_FAIL("Unsupported backends in DirectOperator.Run() with CUDA stream");
+  }
+
+  // Set shared thread pool used for all direct operators.
+  DLL_PUBLIC inline static void SetThreadPool(int num_threads, int device_id, bool set_affinity) {
+    shared_thread_pool = std::make_unique<ThreadPool>(num_threads, device_id, set_affinity);
+  }
+
+  // Set shared CUDA stream used for all direct operators.
+  DLL_PUBLIC inline static void SetCudaStream(int device_id) {
+    if (device_id != CPU_ONLY_DEVICE_ID) {
+      DeviceGuard g(device_id);
+      shared_cuda_stream = CUDAStreamPool::instance().Get(device_id);
+    }
+  }
+
+ private:
+  template <typename InBackend, typename OutBackend, typename WSInputType, typename WSOutputType>
+  std::vector<std::shared_ptr<TensorList<OutBackend>>> RunImpl(
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs);
+
+  int batch_size;
+  size_t num_outputs;
+  workspace_t<Backend> ws;
+  OpSpec op_spec;
+  std::unique_ptr<OperatorBase> op;
+
+  static cudaStream_t shared_cuda_stream;
+  static std::unique_ptr<ThreadPool> shared_thread_pool;
+};
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<CPUBackend>>> DirectOperator<CPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    ThreadPool *thread_pool) {
+  ws.Clear();
+  ws.SetThreadPool(thread_pool);
+
+  return RunImpl<CPUBackend, CPUBackend, TensorVector<CPUBackend>, TensorVector<CPUBackend>>(
+      inputs, kwargs);
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<GPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<GPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    cudaStream_t cuda_stream) {
+  ws.Clear();
+  ws.set_stream(cuda_stream);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
+  auto output = RunImpl<GPUBackend, GPUBackend, TensorList<GPUBackend>, TensorList<GPUBackend>>(
+      inputs, kwargs);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
```

Comment:
We could get rid of this one, too - we could (and should) expose associated stream in TensorXxxGPU in Python and just tell the user that the data is available for that stream. We can (and already do) synchronize D2H copies.

Reply:
Creating API for stream exposure seems like a good idea for a follow-up as it'll probably generate a lot of code and is not really that necessary for the debug mode. But fully agree with support for it in terms of eager operators.